### PR TITLE
feat: support functions and custom containers

### PR DESCRIPTION
**Pull request by jyecusch**
I kept the container and functions separate throughout the code, they don't need to be separate once the images are built, but I predict they'll deviate in future, so decided to keep it that way to avoid another large refactor later to split them. Thoughts?

### CODE REVIEW

File: packages/common/src/resources/container-image.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	password: pulumi.Input<string>;
+	imageName: pulumi.Input<string>;
+	server: pulumi.Input<string>;
+	nitricProvider: pulumi.Input<MembraneProviders>;
+}
+
+/**
+ * Image deployment for a custom container in a Nitric project
+ */
+export class NitricContainerImage extends pulumi.ComponentResource {
```
Comment (tjholm) [inline packages/common/src/resources/container-image.ts:40]: Could this share a resource with the function image deployment if the container images are already being built beforehand? This resource will trigger the build a second time (although it should be cached at that point).

File: packages/common/src/types/container.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export interface NitricContainerTriggers {
+	topics?: string[];
+}
```
Comment (tjholm) [inline packages/common/src/types/container.ts]: does this need to be different from the function triggers?

File: packages/plugins/do/src/commands/deploy/do.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ export default class DoDeploy extends BaseCommand {
 		...BaseCommand.flags,
 		containerRegistry: flags.string({
 			char: 'c',
-			description: 'Digital Ocean Container Registry to deploy services to',
+			description: 'Digital Ocean Container Registry to deploy lambdas to',
```
Comment (tjholm) [inline packages/plugins/do/src/commands/deploy/do.ts]: lambdas -> containers?

File: packages/plugins/do/src/tasks/deploy/container.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+				// TODO: Apply docker deployed repository here...
+				repository: image.name,
+				tag: image.imageDigest,
+			},
+			routes: Object.entries(entrypoint.paths)
+				.filter(([, opts]) => opts.target === name)
+				.map(([path]) => ({ path })),
+		},
+	};
+}
```
Comment (tjholm) [inline packages/plugins/do/src/tasks/deploy/container.ts]: does this need to be different from createFunctionServiceSpec?

**Review by tjholm — COMMENTED** [review]
Looks good, I think there are some pulumi resources that could be unified to help de-duplicate some of the function/container code.

File: packages/common/src/resources/container-image.ts (hunk 1 above)
Comment (jyecusch) [reply packages/common/src/resources/container-image.ts:40]: The images for custom containers aren't built ahead of time, just the buildpacks images. I just tried to unify these two, but it feels like more work than it's worth. I think it's the downstream types that are easier to unify than these two.

File: packages/common/src/types/container.ts (hunk 2 above)
Comment (jyecusch) [reply packages/common/src/types/container.ts]: Nope, unified

File: packages/plugins/do/src/commands/deploy/do.ts (hunk 3 above)
Comment (jyecusch) [reply packages/plugins/do/src/commands/deploy/do.ts]: Oops, overzealous WebStorm refactor powers lol. Fixed.

File: packages/plugins/do/src/tasks/deploy/container.ts (hunk 4 above)
Comment (jyecusch) [reply packages/plugins/do/src/tasks/deploy/container.ts]: unified.

File: packages/common/src/resources/container-image.ts (hunk 1 above)
Comment (tjholm) [reply packages/common/src/resources/container-image.ts:40]: done

**Pull request merged**